// @splitsoftware/splitio/es/sync/SegmentUpdateWorker/SegmentUpdateWorker.js
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _bindInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/bind";
import _classCallCheck from "@babel/runtime-corejs3/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/esm/createClass";
import Backoff from '../../utils/backoff';
/**
 * SegmentUpdateWorker class
 */

var SegmentUpdateWorker = /*#__PURE__*/function () {
  /**
   * @param {Object} segmentsStorage
   * @param {Object} segmentsProducer
   */
  function SegmentUpdateWorker(segmentsStorage, segmentsProducer) {
    var _context, _context2;

    _classCallCheck(this, SegmentUpdateWorker);

    this.segmentsStorage = segmentsStorage;
    this.segmentsProducer = segmentsProducer;
    this.maxChangeNumbers = {};
    this.put = _bindInstanceProperty(_context = this.put).call(_context, this);
    this.__handleSegmentUpdateCall = _bindInstanceProperty(_context2 = this.__handleSegmentUpdateCall).call(_context2, this);
    this.backoff = new Backoff(this.__handleSegmentUpdateCall);
  } // Private method
  // Preconditions: this.segmentsProducer.isSynchronizingSegments === false


  _createClass(SegmentUpdateWorker, [{
    key: "__handleSegmentUpdateCall",
    value: function __handleSegmentUpdateCall() {
      var _context3,
          _this = this;

      var segmentsToFetch = _filterInstanceProperty(_context3 = _Object$keys(this.maxChangeNumbers)).call(_context3, function (segmentName) {
        return _this.maxChangeNumbers[segmentName] > _this.segmentsStorage.getChangeNumber(segmentName);
      });

      if (segmentsToFetch.length > 0) {
        this.handleNewEvent = false;
        this.segmentsProducer.synchronizeSegment(segmentsToFetch).then(function () {
          if (_this.handleNewEvent) {
            _this.__handleSegmentUpdateCall();
          } else {
            _this.backoff.scheduleCall();
          }
        });
      }
    }
    /**
     * Invoked by NotificationProcessor on SEGMENT_UPDATE event
     *
     * @param {number} changeNumber change number of the SEGMENT_UPDATE notification
     * @param {string} segmentName segment name of the SEGMENT_UPDATE notification
     */

  }, {
    key: "put",
    value: function put(changeNumber, segmentName) {
      var currentChangeNumber = this.segmentsStorage.getChangeNumber(segmentName);
      if (changeNumber <= currentChangeNumber || changeNumber <= this.maxChangeNumbers[segmentName]) return;
      this.maxChangeNumbers[segmentName] = changeNumber;
      this.handleNewEvent = true;
      this.backoff.reset();
      if (this.segmentsProducer.isSynchronizingSegments()) return;

      this.__handleSegmentUpdateCall();
    }
  }]);

  return SegmentUpdateWorker;
}();

export { SegmentUpdateWorker as default };
