// @splitsoftware/splitio/lib/__tests__/browserSuites/push-refresh-token.spec.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

require("core-js/modules/es.regexp.constructor");

require("core-js/modules/es.regexp.exec");

require("core-js/modules/es.regexp.to-string");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.testRefreshToken = testRefreshToken;

var _now = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/date/now"));

var _setTimeout2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set-timeout"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _splitchangesSince = _interopRequireDefault(require("../mocks/splitchanges.since.-1.json"));

var _splitchangesSince2 = _interopRequireDefault(require("../mocks/splitchanges.since.1457552620999.json"));

var _mysegmentsNicolasSplitIo = _interopRequireDefault(require("../mocks/mysegments.nicolas@split.io.json"));

var _authPushEnabledNicolasSplitIo601secs = _interopRequireDefault(require("../mocks/auth.pushEnabled.nicolas@split.io.601secs.json"));

var _utils = require("../utils");

var _eventSourceMock = _interopRequireWildcard(require("../../sync/__tests__/mocks/eventSourceMock"));

var _index = require("../../index");

var _settings = _interopRequireDefault(require("../../utils/settings"));

// Replace original EventSource with mock
window.EventSource = _eventSourceMock.default;
var userKey = 'nicolas@split.io';
var baseUrls = {
  sdk: 'https://sdk.push-synchronization/api',
  events: 'https://events.push-synchronization/api',
  auth: 'https://auth.push-synchronization/api'
};
var config = {
  core: {
    authorizationKey: '<fake-token-push-1>',
    key: userKey
  },
  urls: baseUrls,
  streamingEnabled: true // debug: true,

};
var settings = (0, _settings.default)(config);
var MILLIS_SSE_OPEN = 100;
var MILLIS_REFRESH_TOKEN = 1000;
/**
 * Sequence of calls:
 *  0.0 secs: initial SyncAll (/splitChanges, /segmentChanges/*), auth, SSE connection -> refresh token scheduled in 1 second.
 *  0.1 secs: SSE connection opened -> syncAll (/splitChanges, /segmentChanges/*)
 *  1.0 secs: refresh-token: reauth, SSE connection
 *  1.2 secs: SSE connection reopened -> syncAll (/splitChanges, /segmentChanges/*)
 */

function testRefreshToken(fetchMock, assert) {
  fetchMock.reset();
  var start, splitio, client; // mock SSE open and message events

  (0, _eventSourceMock.setMockListener)(function (eventSourceInstance) {
    var _context;

    var expectedSSEurl = (0, _concat.default)(_context = "".concat(settings.url('/sse'), "?channels=NzM2MDI5Mzc0_NDEzMjQ1MzA0Nw%3D%3D_NTcwOTc3MDQx_mySegments,NzM2MDI5Mzc0_NDEzMjQ1MzA0Nw%3D%3D_splits,%5B%3Foccupancy%3Dmetrics.publishers%5Dcontrol_pri,%5B%3Foccupancy%3Dmetrics.publishers%5Dcontrol_sec&accessToken=")).call(_context, _authPushEnabledNicolasSplitIo601secs.default.token, "&v=1.1&heartbeats=true");
    assert.equals(eventSourceInstance.url, expectedSSEurl, 'EventSource URL is the expected');
    (0, _setTimeout2.default)(function () {
      eventSourceInstance.emitOpen();
    }, MILLIS_SSE_OPEN); // open SSE connection after 0.1 seconds
  }); // initial sync

  fetchMock.getOnce(settings.url('/splitChanges?since=-1'), {
    status: 200,
    body: _splitchangesSince.default
  });
  fetchMock.getOnce(settings.url('/mySegments/nicolas@split.io'), {
    status: 200,
    body: _mysegmentsNicolasSplitIo.default
  }); // first auth

  fetchMock.getOnce(settings.url("/auth?users=".concat(encodeURIComponent(userKey))), function (url, opts) {
    if (!opts.headers['Authorization']) assert.fail('`/auth` request must include `Authorization` header');
    assert.pass('auth success');
    return {
      status: 200,
      body: _authPushEnabledNicolasSplitIo601secs.default
    };
  }); // sync after SSE opened

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), {
    status: 200,
    body: _splitchangesSince2.default
  });
  fetchMock.getOnce(settings.url('/mySegments/nicolas@split.io'), {
    status: 200,
    body: _mysegmentsNicolasSplitIo.default
  }); // re-auth due to refresh token

  fetchMock.getOnce(settings.url("/auth?users=".concat(encodeURIComponent(userKey))), function (url, opts) {
    var lapse = (0, _now.default)() - start;
    assert.true((0, _utils.nearlyEqual)(lapse, MILLIS_REFRESH_TOKEN), 'reauthentication for token refresh');
    if (!opts.headers['Authorization']) assert.fail('`/auth` request must include `Authorization` header');
    return {
      status: 200,
      body: _authPushEnabledNicolasSplitIo601secs.default
    };
  }); // sync after SSE reopened

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), function () {
    var lapse = (0, _now.default)() - start;
    assert.true((0, _utils.nearlyEqual)(lapse, MILLIS_REFRESH_TOKEN + MILLIS_SSE_OPEN), 'sync after SSE connection is reopened');
  });
  fetchMock.getOnce(settings.url('/mySegments/nicolas@split.io'), {
    status: 200,
    body: _mysegmentsNicolasSplitIo.default
  }); // second re-auth due to refresh token

  fetchMock.getOnce(settings.url("/auth?users=".concat(encodeURIComponent(userKey))), function (url, opts) {
    var lapse = (0, _now.default)() - start;
    assert.true((0, _utils.nearlyEqual)(lapse, MILLIS_REFRESH_TOKEN * 2), 'second reauthentication for token refresh');
    if (!opts.headers['Authorization']) assert.fail('`/auth` request must include `Authorization` header');
    return {
      status: 200,
      body: _authPushEnabledNicolasSplitIo601secs.default
    };
  }); // split sync after SSE reopened

  fetchMock.getOnce(settings.url('/splitChanges?since=1457552620999'), function () {
    var lapse = (0, _now.default)() - start;
    assert.true((0, _utils.nearlyEqual)(lapse, MILLIS_REFRESH_TOKEN * 2 + MILLIS_SSE_OPEN), 'sync after SSE connection is reopened a second time');
    client.destroy().then(function () {
      assert.end();
    });
  });
  fetchMock.getOnce(settings.url('/mySegments/nicolas@split.io'), {
    status: 200,
    body: _mysegmentsNicolasSplitIo.default
  });
  fetchMock.get(new RegExp('.*'), function (url) {
    assert.fail('unexpected GET request with url: ' + url);
  });
  start = (0, _now.default)();
  splitio = (0, _index.SplitFactory)(config);
  client = splitio.client();
}
